contracts/assetsPool/AssetPoolUpgradeable.sol

```diff
@@ -0,0 +1,154 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+import "../interfaces/LiquidityPoolInterfaces.sol";
+import "./proxy/Proxy.sol";
+import "./token/CRToken.sol";
+import "@openzeppelin/contracts-upgradeable/security/ReentrancyGuardUpgradeable.sol";
+import "@uniswap/lib/contracts/libraries/TransferHelper.sol";
+
+/**
+ * @author CRUIZE.
+ * @title Cruize AssetPool.
+ */
+contract AssetPoolUpgradeable is
+    Proxy,
+    OwnableUpgradeable,
+    ReentrancyGuardUpgradeable
+{
+    mapping(address => address) public lpTokens;
+    // change WETH address when you use different test netwrok CURRENT WETH KOVEN network address.
+    address private WETH = 0xd0A1E359811322d97991E03f863a0C30C2cF029C;
+    event WithdrawEvent(address indexed account, uint256 amount);
+    event DepositEvent(address indexed account, uint256 amount);
+    event CreateToken(
+        address indexed tokenAddress,
+        string tokenName,
+        string tokenSymbol,
+        uint8 decimal
+    );
+
+    /*    ************* PROXY CODE START ****************  */
+    address private crContract;
+
+    /*
+     * @function  initialize        will initialize the crContract.
+     * @param     crContract_       Address -  ERC20Upgradeable Contract.
+     */
+
+    function initialize(address crContract_) external initializer {
+        crContract = crContract_;
+        __Ownable_init();
+        __ReentrancyGuard_init();
+    }
+
+    /*
+     * @function createToken  will Clone CRTokenUpgradeable (ERC20 token).
+     * @param    name_       name of   ERC20Upgradeable Contract.
+     * @param    symbol_     symbol of   ERC20Upgradeable Contract.
+     * @param    decimal_    decimal value of ERC20Upgradeable Contract.
+     * @dev incase of ETH  , we will take token address of WETH becuase ETH is a natvie token of Ethereum so we don't have token address of that.
+     */
+
+    function createToken(
+        string memory name,
+        string memory symbol,
+        address tokenAddress,
+        uint8 decimal
+    ) external onlyOwner nonReentrant {
+        require(
+            lpTokens[tokenAddress] == address(0),
+            "1: Cr Token is already created for this WToken."
+        );
+        CRTokenUpgradeable token = CRTokenUpgradeable(createClone(crContract));
+        token.initialize(name, symbol, decimal);
+        lpTokens[tokenAddress] = address(token);
+        emit CreateToken(address(token), name, symbol, decimal);
+    }
+
+    /*    ************* PROXY CODE END ****************  */
+
+    /*
+     * @function  depositeAsset  will deposit user's asset to CRUIZE.
+     * @param     amount          number of token to be deposit.
+     * @param     reserve         token address to deposit.
```

Comment:
Any reason why we're calling it "**reserve**" and not just **token_address** or just **token** ? Just curious, is using **reserve** a common term for liquidity pool contracts in solidity ?

Reply:
And I see that we're already using the word **token** for token addresses in withdraw asset function.

Reply:
yup, so the reserve I saw some of the liquidity contracts like uni swap and Avve they use this naming convention.